test/liquid_voting/participants_test.exs

```diff
@@ -61,8 +71,8 @@ defmodule LiquidVoting.ParticipantsTest do
       assert participant.name == @valid_attrs[:name]
     end
 
-    test "create_participant/1 with valid data creates a uuid" do
-      assert {:ok, %Participant{} = participant} = Voting.create_participant(@valid_attrs)
+    test "create_participant/1 with valid initial data creates a uuid" do
```

Comment:
Not sure `initial` is necessary here, I think it's implied by the function's _function_ 😉 ?

Reply:
I have created @valid_init_data, which has no uuid, for testing a uuid is created when participant created (this is different to @valid_data used in other tests (which includes a valid uuid)

Reply:
but I think uuid is not needed in any of the data sets, since it won't ever changed?

Reply:
oh, that's true. I removed uuid from all @..data and it works, so I'll remove my 'special' @valid_init_data

Reply:
This PR switches the api to using uuids, not ids, across the Participants, Votes, Delegations and Results contexts (and related db tables), as primary keys and foreign keys (where present). It covers issues #67 and #68.  

I had 2 issues related to upserting functions and the structs they return (apologies in advance for the lengthy notes):  

For reference (from before this PR's switch to uuids):  
```
def upsert_participant(attrs \\ %{}) do  
  %Participant{}  
  |> Participant.changeset(attrs)  
  |> Repo.insert(  
    on_conflict: {:replace_all_except, [:id]},  
    conflict_target: [:organization_uuid, :email]  
  )  
end  
```
part of `calculate_result!/2`:  
```
%Result{}  
|> Result.changeset(attrs)  
|> Repo.insert!(  
  on_conflict: {:replace_all_except, [:id]},  
  conflict_target: [:organization_uuid, :proposal_url]  
)  
```
(I think) the effect of an attempted `Repo.insert!` that becomes a 'replace fields' operation (an update), when using a primary key of an id type (a sequential integer), is that the db does not return an incremented id, since no new row has been created. Thus the returned struct contains the id of the row that was updated. Our uuids, however, are now autogenerated every time a new data struct is created, thus the struct returned after an upsert that just updates (rather than inserting a new row) does not contain the uuid of the updated row, rather it contains a newly generated uuid. I could be wrong about this, but it certainly seems that this is happening.  

1. Previously, when using id as primary key, `upsert_participant(%{email: <"existing_email">, organization_uuid: <valid_organization_uuid>)` would update an existing participant entry with same email, and a `%Participant{}` with the id of the existing participant would be returned. Now, a `%Participant{}` with a newly autogenerated uuid will be returned. The table row will still be updated and the uuid will not be changed in the table, but the returned uuid does not match the record in the table. Thus, I had to introduce an extra step (which sadly involves an extra db query), in `Resolvers.Voting create_vote/3`, which gets the actual uuid of a participant from the relevant db record, rather than simply using the uuid returned in the `%Participant{}` struct, after the preceeding `upsert_participant/1` call.    

2. I replaced the test in voting_results_test.exs of `calculate_result!/1` - "returns the same result struct for a given proposal_url", with; "inserts data into existing result, for a given proposal_url && organization". The old test failed using uuids. This is because it asserted the id returned by the struct was unchanged when a result for the same proposal was created (calculated) twice. This fails because of the uuid autogeneration vs sequential ids issue, described above. Thus the uuids of the structs would never match. My new test looks at what is and isn't inserted to the table instead. Please check if this actually tests what was intended to be tested. Thanks!  

I suspect there is a more elegant way to avoid this issue of the autogenerated uuids not matching after an upsert (that is an update, not an insert), or that I am misunderstanding what is happening. Ideas are welcome :)

As a side note, `upsert_participant(%{attrs})`, when `attrs` does not contain `:name "some_name`, will overwrite a previously existing name with a value of nil, which seems counter-intuitive. This seems particularly strange when considering that GraphQL(absinthe) requires a name and email for createParticipant, but only an email to upsert a participant when using createVote (which will overwrite a pre-existing name with nil, if no name provided).

Other notes:

In the GraphQL examples in the readme, I have indicated where it is necessary to insert a valid uuid from the preceding query(ies). This, unfortunately means that blindly copy/pasting all the examples will result in errors (if valid uuids are not inserted in the appropriate places).

Linting fails on GitHub with error: "The following files were not formatted: * test/liquid_voting_web/absinthe/mutations/create_participant_test.exs" I don't know why, as my last action before my last commit was `mix format`. Help would be appreciated!

Reply:
The detailed notes are excellent, keep them coming! Helps a lot with the review.

The un-matching uuids doesn't sound right, will have a look. And the README examples should definitely work

Reply:
Yeah, that's the only bit I don't like. My 'fix' adds a db hit too, which feels wrong.

Reply:
Also, for clarity, the readme examples do work, IF the user pastes the correct uuid into each query/mutatioin that requires one.

Reply:
Maybe [Ecto.multi](https://hexdocs.pm/ecto/Ecto.Multi.html) will help with the un-matching uuids?